banzai_nres/tests/test_trace_maker.py

```diff
@@ -0,0 +1,595 @@
+                                                                                                       order=1, lastcoef=coefficients_of_last_fit,
+                                                                                                       direction=direction)
+            assert max_exists
+            assert np.isclose(coeff_guess[0], expected_guess * image.data.shape[0], atol=5, rtol=0)
+
+    def test_trimming_bad_fits_on_way_up(self):
+        length = 12
+        all_coefficients_and_indices_init = [[1, 3], [2, 8], [3, 11]]
+        loop_counter = len(all_coefficients_and_indices_init) + 1
+
+        coef = [length+1]
+        all_coefficients_and_indices = all_coefficients_and_indices_init + [[4] + coef]
+        num_orders_found, trimmed_all_coefficients, done = trace_utils.validate_fit_and_trim_erroneous_fits(coef, all_coefficients_and_indices,
+                                                                                                loop_counter, length,
+                                                                                                maximum_exists=True, done=False,
+                                                                                                direction='up')
+        assert done
+        assert trimmed_all_coefficients == all_coefficients_and_indices_init
+        assert num_orders_found == len(trimmed_all_coefficients)
+
+        coef = [all_coefficients_and_indices_init[-1][1] - 3]
+        all_coefficients_and_indices = all_coefficients_and_indices_init + [[4] + coef]
+        num_orders_found, trimmed_all_coefficients, done = trace_utils.validate_fit_and_trim_erroneous_fits(coef, all_coefficients_and_indices,
+                                                                                                loop_counter, length,
+                                                                                                maximum_exists=True, done=False,
+                                                                                                direction='up')
+        assert done
+        assert trimmed_all_coefficients == all_coefficients_and_indices_init
+        assert num_orders_found == len(trimmed_all_coefficients)
+
+        coef = [all_coefficients_and_indices_init[-1][1] - 3]
+        all_coefficients_and_indices = all_coefficients_and_indices_init + [[4] + coef]
+        num_orders_found, trimmed_all_coefficients, done = trace_utils.validate_fit_and_trim_erroneous_fits(coef, all_coefficients_and_indices,
+                                                                                                loop_counter, length,
+                                                                                                maximum_exists=False, done=False,
+                                                                                                direction='up')
+        assert done
```

Comment:
I'm not sure we should have a done flag in the return. Should that raise an exception? Is there a time when it is not done but we want to continue on?